source/common/network/io_socket_error_impl.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "envoy/api/io_error.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Network {
+
+class IoSocketError : public Api::IoError {
+public:
+  explicit IoSocketError(int sys_errno) : errno_(sys_errno) {}
+
+  ~IoSocketError() override {}
+
+  Api::IoError::IoErrorCode getErrorCode() const override;
+  std::string getErrorDetails() const override;
+
+private:
+  int errno_;
+};
+
+// IoErrorCode::Again is used frequently. Define it to be a singleton to avoid frequent memory
+// allocation of such instance. If this is used, IoHandleCallResult has to be instantiated with
+// deleter deleteIoError() below to avoid deallocating memory for this error.
+class IoSocketEagain : public IoSocketError {
+public:
+  IoSocketEagain() : IoSocketError(EAGAIN){};
+};
+
+inline IoSocketEagain* getIoSocketEagainInstance() {
+  static auto* kInstance = new IoSocketEagain();
+  return kInstance;
+}
```

Comment:
can you move this into the .cc file?

Reply:
The impl should still be in the .cc file -- you can declare it here, maybe as a static method of the impl class.

Reply:
Can't you define in header and implement in cc? Same below.